catalyst_voices/lib/widgets/menu/voices_menu.dart

```diff
@@ -0,0 +1,172 @@
+      icon: item.icon,
+      showDivider: item.showDivider,
+      enabled: item.enabled,
+      menuChildren: (item is SubMenuItem)
+          ? item.children.map(_mapItemToButton).toList()
+          : null,
+      onSelected: (item is SubMenuItem) ? null : onTap,
+    );
+  }
+}
+
+class _MenuButton extends StatelessWidget {
+  final String label;
+  final Icon? icon;
+  final bool showDivider;
+  final bool enabled;
+  final List<Widget>? menuChildren;
+  final ValueChanged<String>? onSelected;
+
+  const _MenuButton({
+    required this.label,
+    this.icon,
+    this.showDivider = false,
+    this.enabled = true,
+    this.menuChildren,
+    this.onSelected,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    final theme = Theme.of(context);
+    final textTheme = theme.textTheme;
+
+    final textStyle = textTheme.bodyMedium?.copyWith(
+      color: enabled ? textTheme.bodySmall?.color : theme.disabledColor,
+    );
+
+    final children = menuChildren;
+    return Wrap(
+      children: [
+        Stack(
+          children: [
+            if (icon != null)
+              Positioned.fill(
+                child: Align(
+                  alignment: Alignment.centerLeft,
+                  child: Padding(
+                    padding: const EdgeInsets.only(left: 10),
+                    child: Icon(
+                      icon!.icon,
+                      color: enabled
+                          ? textTheme.bodySmall?.color
+                          : theme.disabledColor,
+                    ),
+                  ),
+                ),
+              ),
+            if (children != null)
+              const Positioned.fill(
+                child: Align(
+                  alignment: Alignment.centerRight,
+                  child: Padding(
+                    padding: EdgeInsets.only(right: 6),
+                    child: Icon(CatalystVoicesIcons.chevron_right, size: 20),
+                  ),
+                ),
+              ),
+            if (children == null)
+              MenuItemButton(
+                leadingIcon: icon,
+                onPressed: enabled ? (() => onSelected?.call(label)) : null,
+                style: MenuItemButton.styleFrom(iconColor: Colors.transparent),
+                child: Text(
+                  label,
+                  style: textStyle,
+                ),
+              )
+            else
+              SubmenuButton(
+                leadingIcon: icon,
+                menuChildren: children,
+                style: MenuItemButton.styleFrom(iconColor: Colors.transparent),
+                child: Text(
+                  label,
+                  style: textStyle,
+                ),
+              ),
+          ],
+        ),
+        if (showDivider)
+          const Divider(
+            height: 0,
+            indent: 0,
+            thickness: 1,
+          ),
+      ],
+    );
+  }
+}
+
+/// Model representing Menu Item
+class MenuItem {
+  final String label;
```

Comment:
What do you think about adding a generic type T to the menu item that would identify it. Using a label for it is rather uncommon because labels can be dynamic, localized, etc. See for instance dropdown button:

```
/// An item in a menu created by a [DropdownButton].
///
/// The type `T` is the type of the value the entry represents. All the entries
/// in a given menu must represent values with consistent types.
class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
  /// Creates an item for a dropdown menu.
  ///
  /// The [child] argument is required.
  const DropdownMenuItem({
    super.key,
    this.onTap,
    this.value,
    this.enabled = true,
    super.alignment,
    required super.child,
  });
```

Reply:
I'm not strong about using generics but I think that using localized labels is a very bad idea. Some items will be dynamic, imagine that you have a contextual menu with a contact book where you display user names, etc. How are you going to distinguish which user item was tapped? Checking their first or last name? What if they are similar?

There must be some technical identifier for each item which uniquely distinguishes which menu item was selected, just my point is, the label is far from being that. Maybe you can come up with some simpler idea than generics but that's what I accustomed to.

Reply:
That should work, thanks!

Reply:
Cool, done!